### ISSUE TITLE
[Bug]: FLASH_ATTN_MLA_SPARSE emits 'Failed to initialize the TMA descriptor 1' every step for GLM-5.3-Flash (zero-extent descriptors; output correct)

### ISSUE THREAD

**Issue by jtatum** [vllm-project/vllm#56563]
### Your current environment

<details>
<summary>vllm collect-env (abridged)</summary>

```
OS                           : Ubuntu 24.04.3 LTS (x86_64)
PyTorch version              : 2.13.0+cu130
CUDA runtime version         : 13.0.88
GPU                          : 8x NVIDIA H100 80GB HBM3 (single node, TP=8)
Nvidia driver version        : 595.71.05
vLLM Version                 : 0.28.1rc1.dev580+g385dce36b (git sha: 385dce36b)
Docker image                 : vllm/vllm-openai:glm53-flash @ sha256:819ec9c063412e5730d1b0e82046ba540d1bf991f3c4f661a849aae8a0c52374
[pip3] flashinfer-python==0.6.18
[pip3] torch==2.13.0+cu130
[pip3] triton==3.7.1
```
</details>

### 🐛 Describe the bug

**Summary:** With `zai-org/GLM-5.3-Flash` and `--attention-backend FLASH_ATTN_MLA_SPARSE` on H100 (TP=8, MTP speculative decoding), every forward step emits ~88 copies of

```
Error: Failed to initialize the TMA descriptor 1
```

from the FA3 sparse-MLA kernel (`vllm/vllm_flash_attn/_vllm_fa3_C.abi3.so`). The dumped descriptors all have `gmem_address 0` and a zero leading extent, i.e. the kernel is asked to build TMA descriptors for **empty (zero-token) segments**. CUDA rejects them with `CUDA_ERROR_INVALID_VALUE` (1). Generation output is correct (verified on short prompts, a 10k-token needle-count prompt, and greedy essays), so the failure is on an empty segment with nothing to compute — but it is emitted on every step in both prefill and decode, first appears during CUDA-graph capture, and produces roughly 2 GB/hour of container log at light single-user load.

**Count matches layers x ranks:** 854 messages for one 60-token completion (~11 MTP steps) and 283 for a `max_tokens=1` request => ~80-90 per step = 11 sparse-MLA layers x 8 TP ranks. This looks like the same class of bug as the FlashMLA sparse "phantom / zero-query-length chunk" reports (#49922, #51486, #52339; partial fix in #51489), except FA3 prints and continues instead of asserting.

**Descriptor dumps (representative):**

```
format         9            # BFLOAT16
dim            4
gmem_address   0
globalDim      (0,512,1,1,1)
globalStrides  (2,4096,4096,0,0)
boxDim         (64,64,1,1,1)
elementStrides (1,1,1,1,1)
interleave     0
swizzle        3
l2Promotion    2
oobFill        0
Error: Failed to initialize the TMA descriptor 1
```

Distinct `globalDim` values observed over ~100k occurrences (count, shape):
```
36694  (0,1,1,2177280,1)
18866  (0,1,1,327680,1)
  687  (0,480,1,1,1)
  547  (0,6,1,1,1)
```

**Launch command:**

```bash
docker run --gpus all --ipc=host -p 8000:8000 \
  -v /path/to/hf-cache:/root/.cache/huggingface \
  -e VLLM_ENGINE_READY_TIMEOUT_S=3600 \
  vllm/vllm-openai:glm53-flash \
  zai-org/GLM-5.3-Flash \
  --served-model-name zai-org/GLM-5.3-Flash \
  --tensor-parallel-size 8 \
  --attention-backend FLASH_ATTN_MLA_SPARSE \
  --speculative-config '{"method":"mtp","num_speculative_tokens":5}' \
  --tool-call-parser glm47 --enable-auto-tool-choice --reasoning-parser glm45 \
  --max-model-len 262144 --max-num-seqs 512
```

Relevant startup lines:
```
Using FLASH_ATTN_MLA_SPARSE attention backend ...
Fused multi-step draft decode is not supported by attention backend(s) DEEPSEEK_V32_INDEXER, FLASH_ATTN_MLA_SPARSE, KPOOL_TAIL; falling back to rebuilding attention metadata between draft steps.
```

**Reproduce:** start as above, then send any chat completion, e.g.
```bash
curl -s localhost:8000/v1/chat/completions -H 'Content-Type: application/json' \
  -d '{"model":"zai-org/GLM-5.3-Flash","messages":[{"role":"user","content":"Say hi."}],"max_tokens":1}'
docker logs <container> 2>&1 | grep -c "TMA descriptor"   # grows by ~280 for that single request
```

**Not present with the default backend:** the same nightly launched without `--attention-backend` (auto-selects `FLASHINFER_MLA_SPARSE_SM90`) logs 0 of these messages across the same benchmark runs.

**Why we are on this backend anyway:** with `vllm bench serve` the auto-selected `FLASHINFER_MLA_SPARSE_SM90` is ~36% slower at concurrency 1 and ~70% slower at concurrency 8 than `FLASH_ATTN_MLA_SPARSE` on this model/hardware (https://github.com/vllm-project/vllm/issues/56564), so FA3 is the attractive option — but it should skip TMA descriptor creation / kernel launch when a segment has zero tokens, or the metadata builder should not hand it empty segments. Beyond the log volume, ~88 failing `cuTensorMapEncodeTiled` calls plus stderr prints per step are host-side work on the decode critical path, which matters on CPU-constrained hosts.

**Not reproduced on:** an image built locally 2026-08-26 from the `glm-release` branch (base `487ecf187`, FlashInfer 0.6.17), which auto-selected `FLASH_ATTN_MLA_SPARSE` and, as far as we observed, did not emit these messages (not exhaustively checked).

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by safiullah3915** [comment]
Hi @jtatum ! Thanks for the incredibly detailed write up. I'm going to take a look at this and work on a fix.

To help narrow down whether we should filter these zero-token segments upstream in the Python metadata builder or just add an early exit inside the FA3 C++ kernel it would be helpful to see exactly what is being passed down.

Could you grab a quick Python stack trace of the dispatch call during the decode step? Alternatively if you can dump the exact shapes/arguments being passed into the vllm_flash_attn Python wrapper right before the TMA descriptor fails!

**Comment by jtatum** [comment]
Thanks @safiullah3915. I captured the arguments and it turns out this isn't a zero-token problem. My original write-up misread the descriptors, sorry about that. The leading `0` in `globalDim` is the **head dimension**, not a token count.

**Root cause:** GLM-5.3-Flash uses NoPE MLA on its sparse-attention layers (`qk_rope_head_dim: 0`, `mla_use_nope: true`, `kv_lora_rank: 512`). So in `FlashAttnMLASparseImpl.forward_mqa`, `q = q_rope` is `(tokens, heads, 0)` and `k = kv_rows[:, kv_lora_rank:]` is `(rows, 1, 1, 0)`. Both are zero-width tensors with `data_ptr 0x0`. All of the score comes through `q_v`/`v`. FA3 still builds TMA descriptors for Q and K, and `cuTensorMapEncodeTiled` rejects zero-extent, null-address tensors. That's 3 failed descriptors per call (1 for Q, 2 for K), on every sparse layer, every step, on every rank.

The output is still correct, apparently because a zero-width Q·K contributes nothing to the score. But it means the kernel is being launched with invalid descriptors, and it only works because nothing loads through them.

**Captured on the live server** (decode step, TP=8, MTP, 385dce36b; all 64 captured calls across 8 ranks looked the same apart from device and pointers):

```
q             {"shape": [4, 8, 0], "stride": [2048, 256, 1], "dtype": "torch.bfloat16", "device": "cuda:2", "data_ptr": "0x0", "numel": 0, "contiguous": true}
k             {"shape": [1692160, 1, 1, 0], "stride": [512, 0, 0, 1], "dtype": "torch.bfloat16", "device": "cuda:2", "data_ptr": "0x0", "numel": 0, "contiguous": true}
v             {"shape": [1692160, 1, 1, 512], "stride": [512, 512, 512, 1], "dtype": "torch.bfloat16", "device": "cuda:2", "data_ptr": "0x7d5a2a000000", "numel": 866385920, "contiguous": true}
q_v           {"shape": [4, 8, 512], "stride": [512, 2048, 1], "dtype": "torch.bfloat16", "device": "cuda:2", "data_ptr": "0x7d60fe502000", "numel": 16384, "contiguous": false}
cu_seqlens_q  {"shape": [5], "stride": [1], "dtype": "torch.int32", "device": "cuda:2", "data_ptr": "0x7d60fe514c00", "numel": 5, "contiguous": true, "values": [0, 1, 2, 3, 4]}
seqused_k     {"shape": [4], "stride": [1], "dtype": "torch.int32", "device": "cuda:2", "data_ptr": "0x7d6a7fdf1c00", "numel": 4, "contiguous": true, "values": [1, 2, 3, 4]}
max_seqlen_q  1
max_seqlen_k  2176
block_table   {"shape": [4, 2176], "stride": [2176, 1], "dtype": "torch.int32", "device": "cuda:2", "data_ptr": "0x7d60fe50a000", "numel": 8704, "contiguous": true}
softmax_scale 0.0625
causal        True
```

Descriptors from that call:

```
TMA Desc Addr:   0x7ffc725b4d00
format         9
dim            4
gmem_address   0
globalDim      (0,4,1,1,1)
globalStrides  (2,4096,4096,0,0)
boxDim         (64,64,1,1,1)
elementStrides (1,1,1,1,1)
interleave     0
swizzle        3
l2Promotion    2
oobFill        0
Error: Failed to initialize the TMA descriptor 1
TMA Desc Addr:   0x7ffc725b4d00
format         9
dim            4
gmem_address   0
globalDim      (0,1,1,1692160,1)
globalStrides  (2,0,0,1024,0)
boxDim         (64,64,1,1,1)
elementStrides (1,1,1,1,1)
interleave     0
swizzle        3
l2Promotion    2
oobFill        0
Error: Failed to initialize the TMA descriptor 1
TMA Desc Addr:   0x7ffc725b4d00
format         9
dim            4
gmem_address   0
globalDim      (0,1,1,1692160,1)
globalStrides  (2,0,0,1024,0)
boxDim         (64,64,1,1,1)
elementStrides (1,1,1,1,1)
interleave     0
swizzle        3
l2Promotion    2
oobFill        0
Error: Failed to initialize the TMA descriptor 1
```

Python stack (innermost frames). Note that this is the eager attention op between piecewise CUDA-graph segments:

```
File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/breakable_cudagraph.py", line 333, in __call__
File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/breakable_cudagraph.py", line 421, in _replay
File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/breakable_cudagraph.py", line 214, in replay
File "/usr/local/lib/python3.12/dist-packages/vllm/compilation/breakable_cudagraph.py", line 115, in <lambda>
File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/attention/kv_transfer_utils.py", line 40, in wrapper
File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/attention/mla_attention.py", line 1410, in unified_mla_attention_with_output
File "/usr/local/lib/python3.12/dist-packages/vllm/model_executor/layers/attention/mla_attention.py", line 1037, in forward_impl
File "/usr/local/lib/python3.12/dist-packages/vllm/v1/attention/backends/mla/flashattn_mla_sparse.py", line 236, in forward_mqa
File "/usr/local/lib/python3.12/dist-packages/vllm/vllm_flash_attn/flash_attn_interface.py", line 456, in flash_attn_varlen_func
```

**Standalone repro** (no model needed; one H100, runs in a few seconds in `vllm/vllm-openai` built from 385dce36b):

```python
# Minimal repro for vllm#56563: FA3 with a zero-width Q/K head dim (NoPE MLA, qk_rope_head_dim=0).
# Mirrors FlashAttnMLASparseImpl.forward_mqa for GLM-5.3-Flash at TP=8 (8 heads/rank, kv_lora_rank 512).
import os, tempfile, torch
from vllm.vllm_flash_attn.flash_attn_interface import flash_attn_varlen_func

def count_tma_errors(fn):
    saved = os.dup(2); tmp = tempfile.TemporaryFile(); os.dup2(tmp.fileno(), 2)
    try:
        out = fn(); torch.cuda.synchronize()
    finally:
        os.dup2(saved, 2); os.close(saved)
    tmp.seek(0); n = tmp.read().decode().count("Failed to initialize the TMA descriptor"); tmp.close()
    return out, n

torch.manual_seed(0)
dev, dt = "cuda", torch.bfloat16
T, H, L, N, TOPK = 4, 8, 512, 4096, 64       # tokens, heads, kv_lora_rank, cache rows, top-k
scale = 0.0625
q_nope = torch.randn(T, H, L, device=dev, dtype=dt)            # passed as q_v
latent = torch.randn(N, L, device=dev, dtype=dt)               # sparse-MLA cache rows (page size 1)
block_table = torch.stack([torch.randperm(N, device=dev)[:TOPK] for _ in range(T)]).to(torch.int32)
cu_q = torch.arange(T + 1, device=dev, dtype=torch.int32)
seqused_k = torch.full((T,), TOPK, device=dev, dtype=torch.int32)

def call(rope):
    kv = torch.cat([latent, torch.zeros(N, rope, device=dev, dtype=dt)], dim=1)
    return flash_attn_varlen_func(
        q=torch.zeros(T, H, rope, device=dev, dtype=dt),
        k=kv[:, L:].unsqueeze(1).unsqueeze(1),
        v=kv[:, :L].unsqueeze(1).unsqueeze(1),
        q_v=q_nope, max_seqlen_q=1, cu_seqlens_q=cu_q, max_seqlen_k=TOPK,
        seqused_k=seqused_k, block_table=block_table, softmax_scale=scale,
        causal=True, fa_version=3)

out0, n0 = count_tma_errors(lambda: call(0))     # production layout: zero-width rope
out64, n64 = count_tma_errors(lambda: call(64))  # control: rope padded to 64 with zeros

# Reference: NoPE MLA score is q_nope . latent; value is latent.
sel = latent[block_table.long()].float()                              # (T, TOPK, L)
p = torch.softmax(torch.einsum("thl,tkl->thk", q_nope.float(), sel) * scale, dim=-1)
ref = torch.einsum("thk,tkl->thl", p, sel)
err = lambda o: (o.float() - ref).abs().max().item()
print(f"rope=0 : TMA descriptor errors={n0}  max|out-ref|={err(out0):.4f}")
print(f"rope=64: TMA descriptor errors={n64}  max|out-ref|={err(out64):.4f}")
```

Output:

```
rope=0 : TMA descriptor errors=3  max|out-ref|=0.0042
rope=64: TMA descriptor errors=0  max|out-ref|=0.0042
```

**On your question:** there are no zero-token segments to filter in the metadata builder. The token count is real (4 here). What's zero is the Q/K head dimension, for every call on these layers. So the options seem to be:

- FA3 treats `headdim_qk == 0` as "qv-only" and skips the Q/K TMA descriptors and the Q·K GEMM. That looks like the proper fix, since that operand is structurally absent.
- The Python backend avoids handing FA3 zero-width Q/K for NoPE models. Padding the rope width to 64 with zeros silences it in the repro and gives the same output. But in production it would mean a zero K operand spanning every cache row, and I haven't tested that on the server.

I'm happy to test either on the H100 box.

**Comment by jtatum** [comment]
Stand by... this may have been fixed in another PR.

**Comment by jtatum** [comment]
Tested the 9/24 nightly. PR #55385 fixed this issue.
